openpype/lib/events.py

```diff
@@ -17,174 +16,83 @@ class MissingEventSystem(Exception):
     pass
 
 
-# Python 3.4+ supports 'partialmethod'
-# TODO remove when support for Python 2.x is completely dropped
-_partial_types = [functools.partial]
-if hasattr(functools, "partialmethod"):
-    _partial_types.append(functools.partialmethod)
-_partial_types = tuple(_partial_types)
+def _get_func_ref(func):
+    if inspect.ismethod(func):
+        return WeakMethod(func)
+    return weakref.ref(func)
 
 
-class _WrappedFunc(object):
-    """Wrap function to be able to check if reference is valid.
+def _get_func_info(func):
+    path = "<unknown path>"
+    if func is None:
+        return "<unknown>", path
 
-    Can create sub-wrappers for 'partial' functions.
+    if hasattr(func, "__name__"):
+        name = func.__name__
+    else:
+        name = str(func)
 
-    Args:
-        func (Callable): Function to wrap.
-    """
-
-    def __init__(self, func):
-        is_partial = False
-        args = None
-        kwargs = None
-        sub_func = None
-        if isinstance(func, _partial_types):
-            is_partial = True
-            func_ref = None
-            args = func.args
-            kwargs = func.keywords
-            sub_func = _WrappedFunc(func.func)
-            expect_args = False
-            expect_kwargs = False
-        else:
-            # Convert callback into references
-            #   - deleted functions won't cause crashes
-            if inspect.ismethod(func):
-                func_ref = WeakMethod(func)
-            else:
-                func_ref = weakref.ref(func)
-
-            # Get expected arguments from function spec
-            # - positional arguments are always preferred
-            expect_args = is_func_signature_supported(func, "fake")
-            expect_kwargs = is_func_signature_supported(func, event="fake")
-
-        self._func_ref = func_ref
+    # Get path to file and fallback to '<unknown path>' if fails
+    # NOTE This was added because of 'partial' functions which is handled,
+    #   but who knows what else can cause this to fail?
+    try:
+        path = os.path.abspath(inspect.getfile(func))
+    except TypeError:
+        pass
 
-        self._is_partial = is_partial
-        self._sub_func = sub_func
-        self._args = args
-        self._kwargs = kwargs
-        self._ref_is_valid = None
-        self._expect_args = expect_args
-        self._expect_kwargs = expect_kwargs
+    return name, path
 
-        self._name = None
-        self._path = None
 
-        self._fill_func_info(func)
+class weakref_partial:
+    """Partial function with weak reference to the wrapped function.
 
-    @property
-    def name(self):
-        """Get name of function.
+    Can be used as 'functools.partial' but it will store weak reference to
+        function. That means that the function must be stored in memory
+        somewhere. e.g. lambda functions are not supported as valid callback.
```

Comment:
Technically `lambda` functions can be used just fine, they just need to keep a reference around.

```
fn = lambda: foobar()
```
The `fn` should be kept around then.

Similar to:
```
def foobar():
    pass
```

---

```suggestion
    Can be used as 'functools.partial' but it will store weak reference to
        function. That means that the function must be reference counted
        to avoid garbage collecting the function itself. 
        
        When the referenced functions is garbage collected then calling the 
        weakref partial (no matter the args/kwargs passed) will do nothing.
        It will fail silently, returning `None`. The `is_valid()` method can be
        used to detect whether the reference is still valid.
       
```